### PR TITLE
DOCSP-30545: serialization guide

### PR DESCRIPTION
**Pull request by rustagir**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-rust/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-30545>
Staging - https://preview-mongodbrustagir.gatsbyjs.io/rust/DOCSP-30545-serialization/fundamentals/serialization/

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/fundamentals/database-collection.txt [hunk 1]
```diff
@@ -150,23 +150,31 @@ perform the following actions:
...
+data type you want to serialize the collection's
 data into. When you call a method on a ``Collection`` instance that is
 parameterized with a specific type, the method accepts or returns
 instances of this type.
 
+.. note::
+   
+   If you do not parameterize your ``Collection``
+   instance, the collection defaults to using the ``Document`` type for
+   data operations.
```
Comment (isabelatkinson) [inline source/fundamentals/database-collection.txt:168]: This used to be true but no longer is; we removed the default generic type for `Collection` in the 2.0 driver release. Users are required to specify a type for their collection now.

As a side note (I'm not sure if this would actually be worth mentioning in the docs), the compiler can infer the type of the collection if a CRUD method is called with a concrete data type in the same scope. For example, the following will compile:

```
// no type annotation
let collection = client.database("db").collection("coll");
// compiler can infer that collection is a Collection<Document> because a `Document` is being inserted
collection.insert_one(doc! { "x": 1 }, None).await?;
```

**Review by isabelatkinson — COMMENTED** [review]
the serde info looks great, you hit all the points I had in mind! I just have one suggestion regarding collection parameterization

File: source/fundamentals/database-collection.txt (hunk 1 above)
Comment (rustagir) [reply source/fundamentals/database-collection.txt:168]: thanks for the clarification! I think that is worth mentioning, I'll revise that section.

File: source/fundamentals/serialization.txt [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+following ``derive`` statement before defining a Rust type:
+
+.. code-block:: rust
+   
+   #[derive(Serialize, Deserialize)]
+
+Custom Struct Example
+~~~~~~~~~~~~~~~~~~~~~
+
+The following code defines a sample ``Vegetable`` struct that derives the
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:77]: S:
```suggestion
The following code defines a sample ``Vegetable`` struct that implements the
```

File: source/fundamentals/serialization.txt [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+We recommend that you define and use a custom type to model your
+collection's data, instead of the ``Document`` type.
+
+.. _rust-custom-data-model:
+
+Custom Data Model
+-----------------
+
+You can use any Rust data type that implements the ``Serialize`` and
+``Deserialize`` traits from the ``serde`` crate as the generic parameter for a
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/fundamentals/serialization.txt:66]: Applies to other mentions of "generic parameter" as well (unless Rust usually references generic parameters rather than generic type parameters)

File: source/fundamentals/serialization.txt [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+           "item": "lima beans"
+         }
+
+Serialize a DateTime as a String
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You might want to represent a ``DateTime`` field value in a document as
+an ISO-formatted string in BSON. To specify this conversion, use the
+``serialize_bson_datetime_as_rfc3339_string`` helper function as the value of
+the ``serialize_with`` attribute attached to the field. In the following
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:202]:
```suggestion
the ``serialize_with`` attribute attached to the field with a ``DateTime`` value. In the following
```

File: source/includes/fundamentals/code-snippets/serialization.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    radius: f32,
+}
+
+#[tokio::main]
+async fn main() -> mongodb::error::Result<()> {
+    let uri: &str = "<connection string>";
+    let client = Client::with_uri_str(uri).await?;
+
+    // begin-access-coll
+    let my_coll: Collection<Vegetable> = client.database("db").collection("vegetables");
```
Comment (norareidy) [inline source/includes/fundamentals/code-snippets/serialization.rs:31]: S: There's some horizontal scroll for this example, maybe split into two lines

File: source/includes/fundamentals/code-snippets/serialization.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    my_coll.insert_one(calabash, None).await?;
+    // end-insert-veg
+
+    // begin-multiple-types
+    let shapes_coll: Collection<Square> = client.database("db").collection("shapes");
+    // ... perform some operations with Square
+
+    let shapes_coll: Collection<Circle> = shapes_coll.clone_with_type();
+    // ... perform some operations with Circle
```
Comment (norareidy) [inline source/includes/fundamentals/code-snippets/serialization.rs:50]: S: there's also a little horizontal scroll for this example

File: source/fundamentals/serialization.txt [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+the ``serialize_with`` attribute attached to the ``_id`` field:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/serialization.rs
+   :language: rust
+   :dedent:
+   :start-after: begin-hex-to-objectid
+   :end-before: end-hex-to-objectid
+
+Select from the :guilabel:`Struct` and :guilabel:`BSON` tabs below
+to see how the driver serializes a sample ``Order`` struct to BSON:
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:172]: S:

 1. put the second half of this sentence first, i.e.: "To see how the driver serializes a sample ``Order`` struct to BSON, select from ..."

 2. style guide doesn't love "below", you could use "Select from the following Struct and BSON tabs" instead

File: source/fundamentals/serialization.txt [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+example, the ``delivery_date`` field has a value of type ``DateTime``:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/serialization.rs
+   :language: rust
+   :dedent:
+   :start-after: begin-dt-to-string
+   :end-before: end-dt-to-string
+
+Select from the :guilabel:`Struct` and :guilabel:`BSON` tabs below
+to see how the driver serializes a sample ``Order`` struct to BSON:
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:212]: S: same suggestion as above (lines 171-172)

File: source/fundamentals/serialization.txt [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+           "delivery_date": "2023-09-26T17:30:18.181Z"
+         }
+
+Serialize a u32 as an f64
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You might want to represent a ``u32`` field value in a document as
+an ``f64``, or ``Double``, type in BSON. To specify this conversion, use the
+``serialize_u32_as_f64`` helper function as the value of
+the ``serialize_with`` attribute attached to the field. In the following
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:243]:
```suggestion
the ``serialize_with`` attribute attached to the field with a ``u32`` value. In the following
```

File: source/fundamentals/serialization.txt [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+         }
+
+Serialize a DateTime as a String
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You might want to represent a ``DateTime`` field value in a document as
+an ISO-formatted string in BSON. To specify this conversion, use the
+``serialize_bson_datetime_as_rfc3339_string`` helper function as the value of
+the ``serialize_with`` attribute attached to the field. In the following
+example, the ``delivery_date`` field has a value of type ``DateTime``:
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:203]: S: since this code example is showing how to represent DateTime as a string in BSON, it would be helpful for the example's introductory text to include that: 
"In the following example, the `serialize_with` attribute specifies the conversion of the DateTime type to a string type in BSON:" (or something similar)

File: source/fundamentals/serialization.txt [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+         }
+
+Serialize a u32 as an f64
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You might want to represent a ``u32`` field value in a document as
+an ``f64``, or ``Double``, type in BSON. To specify this conversion, use the
+``serialize_u32_as_f64`` helper function as the value of
+the ``serialize_with`` attribute attached to the field. In the following
+example, the ``quantity`` field has a value of type ``u32``:
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:244]: S: same suggestion as above (line 203)

File: source/fundamentals/serialization.txt [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+their functionality:
+
+- ``rename``: serialize and deserialize a field with a specified name instead of the
+  Rust struct or variant name
+- ``skip``: do not serialize or deserialize the specified field
+- ``default``: if no value is present during deserialization, use the
+  default value from ``Default::default()``
+
+To see a full list of ``serde`` attributes, see the `serde Attributes
+API documentation <https://serde.rs/attributes.html>`__
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:288]:
```suggestion
API documentation <https://serde.rs/attributes.html>`__.
```

**Review by norareidy — CHANGES_REQUESTED** [review]
Really good, just added some suggestions

File: source/fundamentals/serialization.txt [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+  and methods mentioned in this guide
+
+.. _rust-generic-param:
+
+Generic Type Parameter
+----------------------
+
+When you create a ``Collection`` instance, you must specify a generic
+type parameter to represent the type of data that models the documents
+in your collection. To learn more about specifying a generic parameter,
```
[the remark this answers is not in the document]
Comment (rustagir) [reply source/fundamentals/serialization.txt:53]: changed in all places

File: source/fundamentals/serialization.txt (hunk 4 above)
Comment (rustagir) [reply source/fundamentals/serialization.txt:202]: changed for examples 2 & 3

File: source/fundamentals/serialization.txt (hunk 7 above)
Comment (rustagir) [reply source/fundamentals/serialization.txt:172]: changed both

File: source/fundamentals/serialization.txt [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+
+You can modify the default serialization and deserialization behavior of
+the {+driver-short+} by using **attributes** from the ``serde`` crate.
+Attributes are optional pieces of metadata attached to fields of
+structs or variants of enums.
+
+The ``serde`` crate provides the ``serialize_with`` and
+``deserialize_with`` attributes to select helper functions to customize
+serialization and deserialization on specific fields and variants. To
+specify an attribute on a struct field or enum variant, include it
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:140]: S: just to clarify what "it" refers to, 

```suggestion
specify an attribute on a struct field or enum variant, include the attribute
```

File: source/fundamentals/serialization.txt [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+--------------------
+
+You can modify the default serialization and deserialization behavior of
+the {+driver-short+} by using **attributes** from the ``serde`` crate.
+Attributes are optional pieces of metadata attached to fields of
+structs or variants of enums.
+
+The ``serde`` crate provides the ``serialize_with`` and
+``deserialize_with`` attributes to select helper functions to customize
+serialization and deserialization on specific fields and variants. To
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:139]: S: not a necessary change but I do find the length/complexity of this sentence a little hard to follow. Maybe break it up, like: 

"The serde crate provides the serialize_with and deserialize_with attributes, which take helper functions as values. These helper functions customize serialization and deserialization on specific fields and variants."

File: source/fundamentals/serialization.txt [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+.. code-block:: rust
+   
+   #[derive(Serialize, Deserialize)]
+   struct MyStruct {
+       #[serde(with = "<module>")]
+       field1: u32,
+       // ... other fields
+   }
+
+To see a full list of these modules, see the `serde_helpers API
```
Comment (norareidy) [inline source/fundamentals/serialization.txt:289]: S: nit, but to prevent using "see" twice:

```suggestion
For a full list of these modules, see the `serde_helpers API
```

**Review by norareidy — APPROVED** [review]
LGTM + small suggestions

File: source/fundamentals/serialization.txt (hunk 16 above)
Comment (norareidy) [reply source/fundamentals/serialization.txt:289]: applies elsewhere too if you do decide to reword

**Review by isabelatkinson — APPROVED** [review]
